File: csharp/Subscription.Api/Services/EmailValidationService.cs
namespace Subscription.Api.Services;

public class EmailValidationService(HttpClient httpClient, ILogger<EmailValidationService> logger) : IEmailValidationService
{
    public async Task<bool> IsEmailSafeAsync(string email)
    {
        try
        {
            logger.LogInformation("Validating email safety for: {Email}", email);
            var response = await httpClient.GetAsync($"/validate?email={Uri.EscapeDataString(email)}");

            if (response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                logger.LogInformation("Email validation response: {Response}", content);

                return content.Contains("\"safe\":true", StringComparison.OrdinalIgnoreCase);
            }

            logger.LogWarning("Email validation service returned status: {StatusCode}", response.StatusCode);
            return false;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Error validating email safety for: {Email}", email);
            return false;
        }
    }
}

File: typescript/src/subscription/email-validation-service.ts
import { Injectable } from '@nestjs/common'

@Injectable()
export class EmailValidationService {
  baseUrl: string = 'https://api.emailvalidation.com'

  public async isEmailSafe(email: string): Promise<boolean> {
    try {
      console.info('Validating email safety for: %s', email)
      const response = await fetch(
        `${this.baseUrl}/validate?email=${encodeURIComponent(email)}`,
      )

      if (response.ok) {
        const content = await response.text()
        console.info('Email validation response: %s', content)

        return content.toLowerCase().includes('"safe":true')
      }

      console.warn(
        'Email validation service returned status: %d',
        response.status,
      )
      return false
    } catch (error) {
      console.error(
        'Error validating email safety for: %s',
        error instanceof Error ? error : new Error(String(error)),
        email,
      )
      return false
    }
  }
}
